sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala

```diff
@@ -1167,4 +1172,31 @@ class DSV2CharVarcharTestSuite extends CharVarcharTestSuite
       }
     }
   }
+
+  test("SPARK-44414: Fixed matching check for CharType/VarcharType") {
+    Seq("{\"__CHAR_VARCHAR_TYPE_STRING\":\"varchar(80)\"}",
+      "{\"__CHAR_VARCHAR_TYPE_STRING\":\"char(80)\"}").foreach(json => {
+
+      // query
+      val rows = Seq(Row("1", "2"))
+      val metadata = new MetadataBuilder().withMetadata(Metadata.fromJson(json)).build()
+      val schema = StructType(Seq(
+        StructField("x", StringType, metadata = metadata),
+        StructField("y", StringType, metadata = metadata)))
```

Comment:
I'm not really sure this is an end-to-end test... `__CHAR_VARCHAR_TYPE_STRING` is an internal metadata key and we don't expect end users to set it.